operator/internal/handlers/lokistack_create_or_update.go

```diff
@@ -416,3 +430,49 @@ func isNamespaceScoped(obj client.Object) bool {
 		return true
 	}
 }
+
+func removeRulesConfigMap(ctx context.Context, req ctrl.Request, c client.Client) error {
+	// Check if the CM exists before proceeding.
+	key := client.ObjectKey{Name: manifests.RulesConfigMapName(req.Name), Namespace: req.Namespace}
+
+	var rulesCm corev1.ConfigMap
+	if err := c.Get(ctx, key, &rulesCm); err != nil {
+		if apierrors.IsNotFound(err) {
+			// resource doesnt exist, so nothing to do.
+			return nil
+		}
+		return kverrors.Wrap(err, "failed to lookup configmap", "name", key)
+	}
+
+	if err := c.Delete(ctx, &rulesCm, &client.DeleteOptions{}); err != nil {
+		return kverrors.Wrap(err, "failed to delete configmap",
+			"name", rulesCm.Name,
+			"namespace", rulesCm.Namespace,
+		)
+	}
+
+	return nil
+}
+
+func removeRuler(ctx context.Context, req ctrl.Request, c client.Client) error {
+	// Check if the Statefulset exists before proceeding.
+	key := client.ObjectKey{Name: manifests.RulerName(req.Name), Namespace: req.Namespace}
+
+	var ruler appsv1.StatefulSet
+	if err := c.Get(ctx, key, &ruler); err != nil {
+		if apierrors.IsNotFound(err) {
+			// resource doesnt exist, so nothing to do.
+			return nil
+		}
+		return kverrors.Wrap(err, "failed to lookup Statefulset", "name", key)
+	}
+
+	if err := c.Delete(ctx, &ruler, &client.DeleteOptions{}); err != nil {
+		return kverrors.Wrap(err, "failed to delete statefulset",
+			"name", ruler.Name,
+			"namespace", ruler.Namespace,
+		)
+	}
+
+	return nil
+}
```

Comment:
To keep some sanity in this file, we usually put helpers like these into the `handlers/internal` package. In addition this ensures that nobody is going to use them anywhere else accidentaly. The effect is that we can keep our manifests package static w/o apiserver calls during code review time.